### PR TITLE
Model Training

### PR DESCRIPTION
**Pull request by jthomas39**
Fixes #160 
---
### **What was changed?**
•	Added Machine Learning Training Pipeline: Created train_model.py script in the inference/ directory for training Random Forest models to predict flight events
•	Added Training Documentation: Created README_TRAINING.md with comprehensive instructions for model training workflow
•	Added Dependency Management: Created requirements.txt for managing Python ML dependencies (scikit-learn, pandas, numpy, joblib)
•	Created Model Storage: Established inference/models/ directory structure for storing trained model files (bestModel.pkl and finalModel.pkl)
Modified/Added Files:
•	inference/train_model.py (new)
•	inference/README_TRAINING.md (new)
•	inference/models/ (new directory)
---
### **Why was it changed?**
 Required an automated training pipeline that could:
•	Process labeled flight data generated from label_generator.py
•	Train a reliable classifier using the 8 flight parameters (altitude, heading, vertical_speed, velocity, roll, pitch, yaw, g_force)
•	Evaluate model performance with standard metrics (accuracy, precision, recall)
•	Persist trained models for inference in production
Now implemented a complete Random Forest training pipeline that:
•	Loads labeled datasets from inference/Data/labeled_flight_data.csv
•	Performs stratified train-test splitting (80/20)
•	Trains a Random Forest classifier with sensible defaults (100 trees, unlimited depth)
•	Evaluates performance with comprehensive metrics and confusion matrix
•	Saves trained models in standardized pickle format for deployment
---
### **How was it changed?**

Key Implementation Details:
1. Main Training Script (inference/train_model.py):
•	Lines 25-52: FlightEventModelTrainer class initialization with feature column definitions
•	Lines 54-82: load_data() method validates required columns and displays class distribution
•	Lines 84-106: prepare_data() performs stratified train-test split maintaining class balance
•	Lines 108-133: train_model() configures and trains Random Forest with 100 estimators
•	Lines 148-188: evaluate_model() generates predictions and computes accuracy, precision, recall metrics with detailed classification report
•	Lines 190-209: save_model() persists both bestModel.pkl and finalModel.pkl
•	Lines 212-246: main() orchestrates the complete training workflow

--

### **Screenshots (if applicable):**

2026-04-20 16:11:20,719 - INFO - ============================================================
2026-04-20 16:11:20,719 - INFO - Flight Event Prediction - Model Training
2026-04-20 16:11:20,719 - INFO - ============================================================
2026-04-20 16:11:20,720 - INFO -
Input data: C:\Users\scorp\source\repos\PilotDataSynchronization\inference\Data\labeled_flight_data.csv
2026-04-20 16:11:20,720 - INFO - Output directory: C:\Users\scorp\source\repos\PilotDataSynchronization\inference\models
2026-04-20 16:11:20,720 - INFO - Loading data from: C:\Users\scorp\source\repos\PilotDataSynchronization\inference\Data\labeled_flight_data.csv
2026-04-20 16:11:20,780 - INFO - Loaded 16604 samples
2026-04-20 16:11:20,781 - INFO -
Class distribution:
event_label
CRUISE           9161
HIGH_SPEED       3424
LOW_SPEED        2640
TAXI              729
NORMAL_FLIGHT     253
TURN_LEFT         219
TURN_RIGHT         97
HIGH_ALTITUDE      79
HIGH_G_FORCE        2
2026-04-20 16:11:20,791 - INFO -
Preparing data with test size: 0.2
2026-04-20 16:11:20,816 - INFO - Training set: 13283 samples
2026-04-20 16:11:20,817 - INFO - Test set: 3321 samples
2026-04-20 16:11:20,817 - INFO -
============================================================
2026-04-20 16:11:20,817 - INFO - Training Random Forest Model
2026-04-20 16:11:20,817 - INFO - ============================================================
2026-04-20 16:11:20,817 - INFO -
Training on 13283 samples...
[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done  34 tasks      | elapsed:    0.1s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:    0.3s finished
2026-04-20 16:11:21,291 - INFO - ✓ Model training completed
2026-04-20 16:11:21,322 - INFO -
Feature Importance:

       Feature  Importance
      altitude    0.188204
      velocity    0.182987
vertical_speed    0.164102
       heading    0.126342
         pitch    0.125305
           yaw    0.096843
          roll    0.067943
       g_force    0.048274
2026-04-20 16:11:21,326 - INFO -
============================================================
2026-04-20 16:11:21,326 - INFO - Model Evaluation
2026-04-20 16:11:21,326 - INFO - ============================================================
[Parallel(n_jobs=8)]: Using backend ThreadingBackend with 8 concurrent workers.
[Parallel(n_jobs=8)]: Done  34 tasks      | elapsed:    0.0s
[Parallel(n_jobs=8)]: Done 100 out of 100 | elapsed:    0.0s finished
2026-04-20 16:11:21,412 - INFO -
Evaluation Metrics:
2026-04-20 16:11:21,412 - INFO -   Accuracy:  0.9991 (99.91%)
2026-04-20 16:11:21,412 - INFO -   Precision: 0.9991 (99.91%)
2026-04-20 16:11:21,413 - INFO -   Recall:    0.9991 (99.91%)
2026-04-20 16:11:21,413 - INFO -
Detailed Classification Report:

               precision    recall  f1-score   support

       CRUISE       1.00      1.00      1.00      1832
HIGH_ALTITUDE       1.00      0.94      0.97        16
   HIGH_SPEED       1.00      1.00      1.00       685
    LOW_SPEED       1.00      1.00      1.00       528
NORMAL_FLIGHT       0.98      1.00      0.99        51
         TAXI       1.00      1.00      1.00       146
TURN_LEFT           0              0           0          0              0     0         44           0
TURN_RIGHT          0              0           1          0              0     0          0          18
2026-04-20 16:11:21,470 - INFO -
Saving trained models:
2026-04-20 16:11:21,513 - INFO -   ✓ Saved: C:\Users\scorp\source\repos\PilotDataSynchronization\inference\models\bestModel.pkl
2026-04-20 16:11:21,556 - INFO -
============================================================
2026-04-20 16:11:21,556 - INFO - Training Complete!
2026-04-20 16:11:21,556 - INFO - ============================================================
2026-04-20 16:11:21,556 - INFO -
Final Metrics:
2026-04-20 16:11:21,556 - INFO -   Accuracy:  0.9991
2026-04-20 16:11:21,557 - INFO -   Precision: 0.9991
2026-04-20 16:11:21,557 - INFO -   Recall:    0.9991

### CODE REVIEW

**Review by vbramhadevi — APPROVED** [review]
Approved.
